pallets/xyk/src/lib.rs

```diff
@@ -11,92 +8,285 @@ mod mock;
 #[cfg(test)]
 mod tests;
 
-#[cfg(feature = "runtime-benchmarks")]
-mod benchmarking;
-
 #[frame_support::pallet]
 pub mod pallet {
-    use frame_support::pallet_prelude::*;
-    use frame_system::pallet_prelude::*;
+    use codec::{Decode, Encode};
+    use frame_support::{dispatch::Parameter, ensure, pallet_prelude::*};
+    use frame_system::{pallet_prelude::*, RawOrigin};
+    use scale_info::TypeInfo;
+    use sp_runtime::traits::{Hash, StaticLookup};
+
+    use core::cmp::Ord;
 
     /// Configure the pallet by specifying the parameters and types on which it depends.
     #[pallet::config]
-    pub trait Config: frame_system::Config {
-        /// Because this pallet emits events, it depends on the runtime's definition of an event.
+    pub trait Config: frame_system::Config + pallet_assets::Config {
         type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+        type AssetId: IsType<<Self as pallet_assets::Config>::AssetId> + Parameter + Ord + Copy;
+        /// Address where dex stores all the funds
+        type DEXAddr: Get<<Self::Lookup as StaticLookup>::Target>;
+        /// Minimum balance for issued LP tokens
+        type LpMinBalance: Get<Self::Balance>;
+        /// Hash for asset ids. It is used to determenistically get LP asset id from 2 asset ids.
+        type AssetIdHash: Hash;
     }
 
+    pub type AssetId<T> = <T as Config>::AssetId;
+    pub type Balance<T> = <T as pallet_assets::Config>::Balance;
+
     #[pallet::pallet]
     #[pallet::generate_store(pub(super) trait Store)]
     pub struct Pallet<T>(_);
 
-    // The pallet's runtime storage items.
-    // https://docs.substrate.io/v3/runtime/storage
+    #[derive(Debug, Clone, Copy, Encode, Decode, TypeInfo)]
+    pub struct Pair<Id, B> {
+        x: (Id, B),
+        y: (Id, B),
+    }
```

Comment:
In most cases `AssetId` is better than `Id` and `Balance` is better than `B`.

You can probably even do this like in events:
```suggestion
    pub struct Pair<T: Config> {
        x: (AssetId<T>, Balance<T>),
        y: (AssetId<T>, Balance<T>),
    }
```

Reply:
There is a weird issue with that. This structure would require `T` to implement `TypeInfo`:
```
error[E0277]: the trait bound `T: TypeInfo` is not satisfied
  --> pallets/xyk/src/lib.rs:48:15
   |
48 |     #[pallet::storage]
   |               ^^^^^^^ the trait `TypeInfo` is not implemented for `T`
   |
note: required because of the requirements on the impl of `TypeInfo` for `pallet::Pair<T>`
  --> pallets/xyk/src/lib.rs:42:50
   |
42 |     #[derive(Debug, Clone, Copy, Encode, Decode, TypeInfo)]
   |                                                  ^^^^^^^^
43 |     pub struct Pair<T: Config> {
   |                ^^^^^^^^^^^^^^^
   = note: required because of the requirements on the impl of `StaticTypeInfo` for `pallet::Pair<T>`
   = note: required because of the requirements on the impl of `StorageEntryMetadataBuilder` for `frame_support::pallet_prelude::StorageMap<_GeneratedPrefixForStorageLPPairs<T>, frame_support::Blake2_128Concat, <T as pallet::Config>::AssetId, pallet::Pair<T>>`
```

Reply:
https://docs.rs/scale-info/latest/scale_info/#scale_infoskip_type_params